dbms/src/Interpreters/RowRefs.h

```diff
@@ -72,13 +122,23 @@ class AsofRowRefs
         keyf64,
     };
 
+    AsofRowRefs() {}
+    AsofRowRefs(Type t);
+
     static std::optional<Type> getTypeSize(const IColumn * asof_column, size_t & type_size);
 
-    void insert(Type type, LookupLists &, const IColumn * asof_column, const Block * block, size_t row_num);
-    const RowRef * findAsof(Type type, const LookupLists &, const IColumn * asof_column, size_t row_num) const;
+    // This will be synchronized by the rwlock mutex in Join.h
+    void insert(Type type, const IColumn * asof_column, const Block * block, size_t row_num);
+
+    // This will internally synchronize
+    const RowRef * findAsof(Type type, const IColumn * asof_column, size_t row_num) const;
 
 private:
-    Lookups * lookups = nullptr;
+    // Lookups can be stored in a HashTable because it is memmovable
+    // A std::variant contains a currently active type id (memmovable), together with a union of the types
```

Comment:
Just a thought. We could make it even faster if we do not save type in the variant. Both insert() and findAsof() have type as parameter. So the only place we do not know it is dtor. But I do not see an easy way to make it without making the code worse.

Reply:
It sounds like quite a micro-optimisation - if it turns out that with profiling on real workloads that this matters we could look into it to do some further optimisation. 

At the present (without having measured yet, so take with grain of salt), I think the thing that the biggest opportunity for further optimisation is in std::upper_bound. It could be replaced by something that doesn't always have to do the full log(N) binary search:
- Utilising hints where to start looking (based on the assumption that the joinBlocks will be supplied roughly in order of the timestamps and the timestamp is the final column in the primary key of the table, meaning that if you need entry X, you are likely to need entry X+n afterwards for 0 <= n <= small). Not sure what the runtime complexity of such an approach would be as it depends on the exact implementation.

- Using an interpolation search rather than a binary search (http://www.keithschwarz.com/interesting/code/?dir=interpolation-search), which would get you to expected O(log(log(N))) rather than log(N), but under the assumption that the timestamps are uniformly distributed, which I think is fair in most cases, at least when looking at the first (few) iterations of the search.

Or some combination of both.